File: src/chain.ml
let chain_to_machine = Unix.openfile "chain_to_machine" [Unix.O_WRONLY] 0o666
let () = print_endline "Opened chain_to_machine"
let machine_to_chain = Unix.openfile "machine_to_chain" [Unix.O_RDONLY] 0o666
let () = print_endline "Opened machine_to_chain"

let rec run n =
  let b = Bytes.create 8 in
  let b_len = Bytes.length b in
  Bytes.set_int64_ne b 0 n;
  let _write_result = Unix.write chain_to_machine b 0 b_len in
  Format.printf "Wrote int %Ld to machine\n%!" n;
  let _read_result = Unix.read machine_to_chain b 0 b_len in
  Format.printf "Read %d bytes from machine\n" _read_result;
  let n = Bytes.get_int64_ne b 0 in
  Format.printf "Received int %Ld from machine\n%!" n;
  Unix.sleep 1;
  run n
let ()  = run 0L

File: src/chain.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define CHAIN_TO_MACHINE "chain_to_machine"
#define MACHINE_TO_CHAIN "machine_to_chain"

int main(void)
{
    int write_result, read_result;
    int chain_to_machine, machine_to_chain;
    int counter = 0;

    mknod(CHAIN_TO_MACHINE, S_IFIFO | 0666, 0);
    mknod(MACHINE_TO_CHAIN, S_IFIFO | 0666, 0);

    chain_to_machine = open(CHAIN_TO_MACHINE, O_WRONLY);
    printf("opened chain to machine\n");
    machine_to_chain = open(MACHINE_TO_CHAIN, O_RDONLY);
    printf("opened machine to chain\n");

    while (1)
    {
        if ((write_result = write(chain_to_machine, &counter, sizeof(counter))) == -1)
            perror("write");
        else
            printf("wrote counter %d\n", counter);

        if ((read_result = read(machine_to_chain, &counter, sizeof(counter))) == -1)
            perror("read");
        else
        {
            printf("read counter %d\n", counter);
        }
        sleep(1);
    }

    return 0;
}
